File: solutions/4180-pythagorean-distance-nodes-in-a-tree/pythagorean-distance-nodes-in-a-tree.java
class Solution {
    int[] h, ne, to, q;
    int idx;

    void add(int u, int v) {
        to[idx] = v; ne[idx] = h[u]; h[u] = idx++;
        to[idx] = u; ne[idx] = h[v]; h[v] = idx++;
    }

    int[] bfs(int s, int n) {
        int[] d = new int[n];
        for (int i = 0; i < n; i++) d[i] = -1;
        d[s] = 0;
        int head = 0, tail = 0;
        q[tail++] = s;
        while (head < tail) {
            int u = q[head++];
            for (int i = h[u]; i != -1; i = ne[i]) {
                int v = to[i];
                if (d[v] == -1) {
                    d[v] = d[u] + 1;
                    q[tail++] = v;
                }
            }
        }
        return d;
    }

    public int specialNodes(int n, int[][] edges, int x, int y, int z) {
        h = new int[n];
        for (int i = 0; i < n; i++) h[i] = -1;
        ne = new int[2 * n];
        to = new int[2 * n];
        q = new int[n];
        idx = 0;
        for (int[] e : edges) add(e[0], e[1]);

        int[] dx = bfs(x, n), dy = bfs(y, n), dz = bfs(z, n);
        int ans = 0;
        for (int i = 0; i < n; i++) {
            long a = dx[i], b = dy[i], c = dz[i];
            long m = a > b ? (a > c ? a : c) : (b > c ? b : c);
            if (a * a + b * b + c * c == 2 * m * m) ans++;
        }
        return ans;
    }
}

File: solutions/4180-pythagorean-distance-nodes-in-a-tree/pythagorean-distance-nodes-in-a-tree.cpp
class Solution {
public:
    int specialNodes(int n, vector<vector<int>>& edges, int x, int y, int z) {
        vector<vector<int>> g(n);
        g.reserve(n);
        for (auto& e : edges) {
            g[e[0]].push_back(e[1]);
            g[e[1]].push_back(e[0]);
        }

        auto bfs = [&](int start) {
            vector<int> d(n, -1);
            d[start] = 0;
            vector<int> q;
            q.reserve(n);
            q.push_back(start);
            int h = 0;
            while (h < q.size()) {
                int u = q[h++];
                for (int v : g[u]) {
                    if (d[v] == -1) {
                        d[v] = d[u] + 1;
                        q.push_back(v);
                    }
                }
            }
            return d;
        };

        auto dx = bfs(x);
        auto dy = bfs(y);
        auto dz = bfs(z);

        int cc = 0;
        for (int i = 0; i < n; ++i) {
            long long a = dx[i], b = dy[i], c = dz[i];
            long long m = max({a, b, c});
            if (a * a + b * b + c * c == 2 * m * m)  
                cc++;
        }
        return cc;
    }
};